![CC-BY-SA](https://mirrors.creativecommons.org/presskit/buttons/88x31/svg/by-sa.svg)


This notebook was created by [Bernardo Freitas Paulo da Costa](http://www.im.ufrj.br/bernardofpc),
and is licensed under Creative Commons BY-SA.

Antes de enviar este Teste, verifique que tudo está funcionando como esperado.
Por exemplo, **rode o código inteiro, do zero**.
Para isso, vá no menu, escolha _Kernel_, depois _Restart & Run All_.

Verifique, também, que você respondeu todas as questões:
* as questões de código têm `YOUR CODE HERE` (e você pode apagar o `raise NotImplemented` ao incluir sua resposta)
* as questões discursivas têm "YOUR ANSWER HERE".

---

In [1]:
import numpy as np

# Teste 1: Bisseção

## Questão 1: Número de iterações da bisseção, e de chamadas à função `f`

Generalize o algoritmo da bisseção _iterativa_ para retornar,
- a (aproximação da) raiz,
- o número de bisseções (divisões por 2 do intervalo) feitas,
- o número de vezes que você chamou a função `f`

Use como critério de parada
- `xtol`, e
- `maxiter`,

ou seja, o algoritmo termina quando:
- seja possível garantir que o erro (absoluto) da resposta ("em $x$") seja menor do que `xtol`; OU
- quando o algoritmo já tiver feito `maxiter` itererações.

In [2]:
def criterio_parada(a,b,xtol,niter,maxiter): 
    return abs(a - b) < xtol or niter == maxiter
def bissecao(f, a, b, xtol=1e-8, maxiter=10):
    num_iter = 0
    #supoe fb > fa
    if(f(a) > f(b)):
        aux = b
        b = a
        a = aux
    fcalls = 2
    while(not criterio_parada(a,b,xtol,num_iter,maxiter)):
        m = (a+b)/2
        if(f(m) > 0):
            b = m
        else:
            a = m   
        num_iter += 1
        fcalls += 1
    return (a+b)/2, num_iter, fcalls

### 1.1: Testando com $\sqrt2$:

In [3]:
def p2(x): return x**2 - 2

In [4]:
raiz, niters, ncalls = bissecao(p2, 0, 2, maxiter=100)
assert abs(raiz - np.sqrt(2)) < 1e-8
assert 25 <= niters <= 30
assert 25 <= ncalls

In [5]:
raiz, niters, ncalls = bissecao(p2, 0, 2, xtol=1e-3)
assert abs(raiz - np.sqrt(2)) < 2e-3
assert 9 <= niters <= 11

In [6]:
raiz, niters, ncalls = bissecao(p2, 0, 2)
assert abs(raiz**2 - 2) > 1e-3
assert niters == 10

### 1.2: Tolerância, número de iterações, respostas, ...

Observe o seguinte código:

In [7]:
for err in np.logspace(-8, 0, num=9):
    r, ni, nc = bissecao(p2, 0, 2, xtol=err)
    print(f"xtol = {err:.1e} --> raiz = {r:.10f}, #iter = {ni:3d}")

xtol = 1.0e-08 --> raiz = 1.4150390625, #iter =  10
xtol = 1.0e-07 --> raiz = 1.4150390625, #iter =  10
xtol = 1.0e-06 --> raiz = 1.4150390625, #iter =  10
xtol = 1.0e-05 --> raiz = 1.4150390625, #iter =  10
xtol = 1.0e-04 --> raiz = 1.4150390625, #iter =  10
xtol = 1.0e-03 --> raiz = 1.4150390625, #iter =  10
xtol = 1.0e-02 --> raiz = 1.4179687500, #iter =   8
xtol = 1.0e-01 --> raiz = 1.4062500000, #iter =   5
xtol = 1.0e+00 --> raiz = 1.2500000000, #iter =   2


Comente

Quanto menor o erro desejado em x, maior o número de iterações necessárias. No entanto como o maxiter por default é 10, o máximo de precisão que conseguimos com 10 iterações foi 1.4150390625. Se notarmos no exemplo abaixo em que a condição de parada é limitada a 1000 iterações e pelo respectivo erro, vemos que a cada 3 ou 4 iterações o erro (xtol) diminui a uma ordem de grandeza (dividido por 10). Isso porque a cada iteração o tamanho do intervalo decai a metade, e quando temos 3 e 4 iterações, temos que o xtol é divido por 2^3 = 8 e 2^4= 16 respectivamente, o que faz sentido. 

Como seria o equivalente se o critério fosse o número de iterações?

In [8]:
for niter in np.arange(2,50):
    r, ni, nc = bissecao(p2, 0, 2,maxiter = niter)
    print(f"raiz = {r:.10f}, #iter = {ni:3d}")

raiz = 1.2500000000, #iter =   2
raiz = 1.3750000000, #iter =   3
raiz = 1.4375000000, #iter =   4
raiz = 1.4062500000, #iter =   5
raiz = 1.4218750000, #iter =   6
raiz = 1.4140625000, #iter =   7
raiz = 1.4179687500, #iter =   8
raiz = 1.4160156250, #iter =   9
raiz = 1.4150390625, #iter =  10
raiz = 1.4145507812, #iter =  11
raiz = 1.4143066406, #iter =  12
raiz = 1.4141845703, #iter =  13
raiz = 1.4142456055, #iter =  14
raiz = 1.4142150879, #iter =  15
raiz = 1.4141998291, #iter =  16
raiz = 1.4142074585, #iter =  17
raiz = 1.4142112732, #iter =  18
raiz = 1.4142131805, #iter =  19
raiz = 1.4142141342, #iter =  20
raiz = 1.4142136574, #iter =  21
raiz = 1.4142134190, #iter =  22
raiz = 1.4142135382, #iter =  23
raiz = 1.4142135978, #iter =  24
raiz = 1.4142135680, #iter =  25
raiz = 1.4142135531, #iter =  26
raiz = 1.4142135605, #iter =  27
raiz = 1.4142135642, #iter =  28
raiz = 1.4142135642, #iter =  28
raiz = 1.4142135642, #iter =  28
raiz = 1.4142135642, #iter =  28
raiz = 1.4

Comente, também.

Conforme o número de iterações aumenta, mais a raiz se aproxima do valor real. No entanto, quando temos 28 iterações o xtol chega ao erro tolerável, de forma que aumentando o número de iterações não modifica o resultado.

### 1.3: Agora, um polinômio um pouco mais complicado:

In [9]:
def p4(x): return x**4 - 3*x + 1

In [10]:
r1, n1, c1 = bissecao(p4, 0, 1, maxiter=100)
assert p4(r1) < 1e-6

In [11]:
r2, n2, c2 = bissecao(p4, 1, 2, maxiter=100)
assert n1 == n2
assert p4(r2) < 1e-6

In [12]:
r3, n3, c2 = bissecao(p4, 1, 3, maxiter=100)
assert n3 == n2+1
assert r3 == r2

### 1.4: Relação entre `ncalls` e `niters`

Muito provavelmente, é possível calcular `ncalls` a partir de `niters`.
Qual a relação, para o seu programa, entre estes valores?
Porquê?

In [13]:
def compute_ncalls(niters):
    return niters + 2 

Antes de iniciar o programa checa-se se de fato f(a) < f(b). Caso contrário os valores de a e b são trocados.

## Questão 2: Bisseção em listas ordenadas

Se desejamos encontrar um número em uma lista, podemos percorrer cada um de seus índices
até achar o número, e retornamos o índice correspondente.

Mas, se a lista `l` possui números em ordem,
é possível usar uma variante da bisseção para encontrar este índice.
Observe que agora desejamos retornar um número **inteiro** (o índice!),
e que índices para listas sempre devem ser inteiros, portanto:
- cuidado ao escolher o "ponto médio"
- mas o critério de parada é "quando achar o índice".

### 2.1: Encontrando um número

Implemente a função `bissect_list(l, v)`, que retorna
- o índice `k` do elemento na lista `l`, que vale `v`; e
- o número de acessos à lista `l`.

Suponha que a lista `l` é não-decrescente (ou seja, `l[i] <= l[i+1]`).

In [14]:
def generate_m(a,b):
    m = (a+b)/2
    return int(np.floor(m)) 
def bissect_list(l, v):
    """Índice  k  na lista  l , crescente, tal que l[k] = v.  None caso não exista."""
    n = len(l)
    naccess = 0
    b = n - 1
    a = 0
    while(a <= b):
        m = generate_m(a,b)
        #print(f"a = {a}, b = {b}")
        naccess += 1
        A_medio = l[m]
        if(A_medio == v):
            return m,naccess
        if(A_medio > v):
            b = m - 1
        else:
            a = m + 1
    return None, naccess

In [15]:
assert bissect_list([], 42) == (None,0)

In [16]:
l = [4]
assert bissect_list(l, 4) == (0,1)

In [17]:
assert bissect_list(l, 5)[0] is None

In [19]:
## funcao que calcula o valor esperado do numero de iteracoes. SUM_i=1_i=log2N i*2ˆ(i-1)
def compute_expected_value(len_vec):
    depth = np.floor(np.log(len_vec)/np.log(2))
    max_access = depth + 1
    missing_values_to_complete_tree = 2**max_access - len_vec -1
    n_access_per_depth = np.arange(1,max_access+1)
    n_elements_per_depth = np.power(2,np.add(n_access_per_depth,-1))
    n_elements_per_depth[-1] -= missing_values_to_complete_tree
    assert np.sum(n_elements_per_depth) == len_vec
    result = np.dot(n_access_per_depth,n_elements_per_depth)
    return result/len_vec

In [22]:
l = [1,3,6,10,15,21,28]
for i,v in enumerate(l):
    idx, n = bissect_list(l, v)
    assert idx == i, f"v={v}, i={i}, idx={idx}"
    print(n)

3
2
3
1
3
2
3


In [23]:
compute_expected_value(len(l))

2.4285714285714284

Qual o maior número de acessos à lista que você precisou para encontrar um valor nela?
Este número depende da posição na lista?  Explique!

O maior número de acessos foi 3. Esse número depende da posição pois ela determina quantas bissecções serão realizadas. Por exemplo, se o valor que buscamos se situa exatamente no meio, logo na primeira vez que efetuarmos a bissecção chegaremos nele, como é o caso do 10 acima. Por outro lado, valores que situam em outras posições necessitam de mais acessos. O máximo de iterações é na ordem de log2(len(l)) e notamos certa simetria na quantidade bisseções em relação a posição na lista. 

### 2.2: Uma lista muito maior

In [20]:
l = np.arange(-4000, 6000)
np.random.seed(1)
valores_testes = np.random.randint(low=-4000, high=6000, size=10)
for v in valores_testes:
    idx, n = bissect_list(l, v)
    assert l[idx] == v
    assert n <= 20
    print(f"v={v:5d}, idx={idx:5d}, n={n:3d}")

v=-3765, idx=  235, n= 12
v= 1192, idx= 5192, n= 13
v=-3095, idx=  905, n= 14
v= 3813, idx= 7813, n= 12
v=-1105, idx= 2895, n= 12
v= 1056, idx= 5056, n= 14
v=-3856, idx=  144, n= 14
v=  225, idx= 4225, n= 14
v= 3751, idx= 7751, n= 14
v= -538, idx= 3462, n= 12


In [21]:
compute_expected_value(len(l))

12.3631

Como o número de acessos varia, nesta outra lista?
E como ele se compara com o tamanho da lista?

O número de acessos variou entre 12 e 14, aproximadamente log_2(10000) = 13. Isso faz sentido pois, de acordo com o resultado acima, o valor esperado do número de acessos a lista é 12,3631. Já para o caso anterior o valor esperado é na ordem de 2,4285. Assim, para uma lista cerca de 1428x maior, o número médio de acessos só foi multiplicado por 4. 

### 3.3: Listas crescentes e decrescentes.

Altere o código anterior, para receber um argumento a mais, `decr`, que indica que a lista é **decrescente**.
Não reordene a lista!

In [24]:
def bissect(l, v, decr=False):
    """Index  k  on a monotonic list  l  such that  l[k] == v.
    Returns  None  if  not present.
    If  l  is decresing, pass  decr=True."""
    n = len(l)
    naccess = 0
    b = n - 1
    a = 0
    while(a <= b):
        m = generate_m(a,b)
        naccess += 1
        A_medio = l[m]
        if(A_medio == v):
            return m,naccess
        if(A_medio > v):
            if(decr):
                a = m + 1
            else:    
                b = m - 1
        else:
            if(decr):
                b = m - 1
            else:
                a = m + 1
    return None, naccess

In [25]:
l = [1,3,6,10,15,21,28,35,36]
assert bissect(l, 15)[0] == 4

In [26]:
l_rev = [36, 35, 28, 21, 15, 10, 6, 3, 1]
assert bissect(l_rev, 15, decr=True)[0] == 4

In [27]:
l_rev = [36, 35, 28, 21, 15, 10, 6, 3, 1]
assert bissect(l_rev, 16, decr=True)[0] == None

In [28]:
l3 = [36, 35, 28, 21, 15, 10, 9, 6, 3, 2, 1]
assert bissect(l3, 15, decr=True)[0] == 4

### 2.4: Bônus

O número de acessos à lista para encontrar um número muda se esta está invertida?
Explique!

Não muda. Podemos enxergar essa lista como uma árvore binária, onde iremos para esquerda ou direita dependendo se o valor v for menor ou maior do que o "nó" (m) atual, respectivamente. Quando mudamos a ordenação, ao invés de ir para esquerda, iremos seguir o ramo da direita. No entanto a "profundidade da árvore" - número de acessos a lista necessários - se mantém. O caso médio também é identico por consequência.  